### PR TITLE
Added Unit Tests

### PR DESCRIPTION
**Pull request by wk9874**
Added unit tests for each connector:

- MOOSE
- FDS
- OpenFOAM
- TensorFlow

For non python applications, these work using unittest.mock to write example files to temporary directories which mimics the various outputs of each piece of software.

To be able to run the tests you will need to do `poetry install --all-extras`

Fixed a number of bugs:
- VectorPostProcessor parser now uses ID in the metric name instead of x, y, z positions - this is required because some Vector PostProcessors report results where xyz change as well as the variable values, meaning metrics with new names would be created at each timestep
- Added option to enable/disable tracking of Vector PP files - this is because in some applications, these files could be hundreds of lines long and generate hundreds of (not very useful) metrics
- Added option to enable/disable tracking of positions (xyz or radius) from Vector PP files - these should be disabled if your VectorPP files report variables at the same position every timestep, as the position metrics will just be constant values
- OpenFOAM connector now removes Zip files after uploading
- Changed zip files to be just 'inputs.zip' instead of 'constant.zip', 'system.zip', 'initial_conditions.zip' - this means that now the user can pull a single file and unpack it to regain the original case dir structure
- Fixed regex in FDS activation times (previously was not matching with any log lines)
- Fixed bug where max and min divergence labels were the wrong way round in FDS metric RegEx
- Fixed bug in FDS header RegEx's where \n's would be included in the metadata uploaded
- Fixed typing on FDS and MOOSE launch() methods to allow workdir or results dir to be strings or paths
- Renamed validation_accuracy, validation_loss to val_accuracy, val_loss in tensorflow Evaluate runs to match the names in Simulation / Epoch runs
- Added updates and bug fixes to AlertValidator
- Fixed bug in TensorVue where if epoch runs were disabled, the early termination feature wouldnt work

### CODE REVIEW

File: simvue_integrations/connectors/fds.py [hunk 1]
```diff
@@ -75,25 +75,29 @@ def _metrics_callback(self, data, meta):
     @mp_file_parser.file_parser
     def _header_metadata(self, input_file: str, **__) -> tuple[dict[str,typing.Any], list[dict[str, typing.Any]]]:
         """Parse metadata from header of FDS stderr"""
+        print("in here")
```
Comment (kzscisoft) [inline simvue_integrations/connectors/fds.py:78]: A testing artifact that can be removed?

File: simvue_integrations/connectors/fds.py [hunk 2]
```diff
@@ -75,25 +75,29 @@ def _metrics_callback(self, data, meta):
...
             "fds.mpi_processes": re.compile(r"^\s*Number of MPI Processes:\s*(\d+)"),
             "fds.mpi_version": re.compile(r"^\s*MPI version:\s*([\d\.]+)"),
-            "fds.mpi_library_version": re.compile(r"^MPI library version:\s*([\w\d\.\s\*\(\)\[\]\-\_]+)"),
+            "fds.mpi_library_version": re.compile(r"^\s*MPI library version:\s*([\w\d\.\s\*\(\)\[\]\-\_][^\n]+)"),
         }
 
         _output_metadata: dict[str, str] = {}
 
         for line in _file_lines:
+            print(line)
```
Comment (kzscisoft) [inline simvue_integrations/connectors/fds.py:95]: Another testing artifact

File: simvue_integrations/connectors/fds.py [hunk 3]
```diff
@@ -75,25 +75,29 @@ def _metrics_callback(self, data, meta):
...
             for key, regex in _components_regex.items():
                 if (search_res := regex.findall(line)):
                     _output_metadata[key] = search_res[0]
+                    
+        print(_output_metadata)
```
Comment (kzscisoft) [inline simvue_integrations/connectors/fds.py:100]: And another! I strongly recommend having git hooks and the hook that forbids print statements, for required statements use logging

File: simvue_integrations/connectors/moose.py [hunk 4]
```diff
@@ -101,13 +101,14 @@ def _vector_postprocessor_parser(
...
-                # Remove 'id' parameter if it exists, since we don't need it
-                csv_data.pop('id', None)
-                
-                if radius := csv_data.pop('radius', None):
-                    metrics.update({f"{vector_name}.{key}.r={radius}":value for (key, value) in csv_data.items()})
-                elif (x := csv_data.pop('x', None)) is not None and (y := csv_data.pop('y', None)) is not None and (z := csv_data.pop('z', None)) is not None:
-                    metrics.update({f"{vector_name}.{key}.x={x}_y={y}_z={z}":value for (key, value) in csv_data.items()})
+                if self.track_vector_positions == False:
+                    csv_data.pop('x', None)
+                    csv_data.pop('y'), None
```
Comment (kzscisoft) [inline simvue_integrations/connectors/moose.py:106]: Looks like a bug `pop('y', None)`?

File: simvue_integrations/connectors/openfoam.py [hunk 5]
```diff
@@ -52,6 +52,7 @@ def _save_directory(self, dir_names: list[str], zip_name: str, file_type : typin
         if self.upload_as_zip:
             zip_file.close()
             self.save_file(os.path.join(self.openfoam_case_dir, zip_name), file_type)
+            os.remove(os.path.join(self.openfoam_case_dir, zip_name))
```
Comment (kzscisoft) [inline simvue_integrations/connectors/openfoam.py:55]: If repeating the same statement consider using a variable:

```python
out_zip = os.path.join(self.openfoam_case_dir, zip_name)
self.save_file(out_zip, file_type)
os.remove(out_zip)
```
just means if there are changes you only have to modify a select few lines.

File: simvue_integrations/connectors/tensorflow.py [hunk 6]
```diff
@@ -433,8 +431,8 @@ def on_test_batch_end(self, batch, logs):
             if self.create_epoch_runs:
                 self.epoch_run.log_metrics(
                     {
-                        "validation_accuracy": logs.get("accuracy"),
-                        "validation_loss": logs.get("loss"),
+                        "val_accuracy": logs.get("accuracy"),
```
Comment (kzscisoft) [inline simvue_integrations/connectors/tensorflow.py:434]: `val` usually means "value", if in this case it means "validation" I would have left this

File: simvue_integrations/extras/validators.py [hunk 7]
```diff
@@ -52,6 +53,12 @@ class AlertValidator(pydantic.BaseModel):
     # For 'is outside range' or 'is inside range'
     range_low: typing.Optional[typing.Union[int, float]] = None
     range_high: typing.Optional[typing.Union[int, float]] = None
+    aggregation: typing.Optional[typing.Literal["average", "sum", "at least one", "all"]] = "average"
+    trigger_abort: typing.Optional[bool] = None
+    
+    class Config:
```
Comment (kzscisoft) [inline simvue_integrations/extras/validators.py:59]: I think this is a v1 API configuration, you should use v2 where possible:
https://docs.pydantic.dev/dev/concepts/config/

File: simvue_integrations/extras/validators.py [hunk 8]
```diff
@@ -65,6 +72,10 @@ def check_rule(cls, v, values):
     def check_metric(cls, v, values):
         return check_input(v, values, "metric", "source", ["metrics"])
     
+    @pydantic.validator("aggregation", always=True)
```
Comment (kzscisoft) [inline simvue_integrations/extras/validators.py:75]: You are definitely using v1 here, this is now `field_validator` (`root_validator` has become `model_validator` as well): https://docs.pydantic.dev/dev/concepts/validators/#field-validators

**Review by kzscisoft — CHANGES_REQUESTED** [review]

File: simvue_integrations/connectors/fds.py (hunk 3 above)
Comment (wk9874) [reply simvue_integrations/connectors/fds.py:100]: Yep we'll add pre-commit hooks when we chat later :)

File: simvue_integrations/connectors/tensorflow.py (hunk 6 above)
Comment (wk9874) [reply simvue_integrations/connectors/tensorflow.py:434]: Fair point, will change all instances of that

**Review by kzscisoft — APPROVED** [review]
